### ISSUE TITLE
unsound type constraints

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#8459]
```typescript
interface I {
	fn: <T>(t: T) => void;
}
interface X {
	x: number;
}
function toI<T extends X>() : I {
	return {
		fn: (t: T) => { // <-- expected a warning from the compiler that the type constraint cannot be enforced
			alert(t.x);
		}
	};
}
function test(i: I): void {
	i.fn({}); // <-- problem unnoticed
}
test(toI<X>()); // displays "undefined"
```

**Comment by malibuzios** [comment]
The example you gave was a bit too complicated for me to understand but I just wanted to mention that generic function type declarations are not currently enforced, in general.

Perhaps until the [implicit cast of generic function type parameters and return type to `any`](https://github.com/Microsoft/TypeScript/issues/8397#issuecomment-215998291) is addressed, an error can be shown for any generic function type in general?

(There's no current error here)
```ts
declare let f1: <T>(x: T) => T; // Possible error/warning: "generic function types are not 
                                // currently supported/checked"
declare let f2: (x: number) => void;

f1 = f2; // or instead error only error on assignment
```
I feel it would be a bit inappropriate to transfer the responsibility to the volunteers who maintain `tslint`.. Basically to report on an internal implementation issue of the compiler, which may be subject to change in the future, and is not actually related to the language or the design itself.

**Comment by malibuzios** [comment]
I've updated my comment based on [this comment](https://github.com/Microsoft/TypeScript/issues/3410#issuecomment-118952458) in #3410:
>When you call a function with a constrained type parameter, that function's type parameter constraint is checked. When you pass a function with a type parameter constraint, the constraint is not checked.

**Comment by aleksey-bykov** [comment]
bump

**Comment by RyanCavanaugh** [comment]
I believe this is the same root cause as #8397 ?

**Comment by mhegazy** [comment]
tracked by https://github.com/Microsoft/TypeScript/issues/5616
